### PR TITLE
update doc; /regions signature; /grids signature

### PR DESCRIPTION
**Pull request by nickmasson**
## Changes
This refactors the `/regions` and `/grids` signatures and expected behavior.

See the Changelog notes.

GeoJSON response object is also augmented with per-features attributes, meant to localize resource identifying values.  This change opens the door to future feature expansion of the `regions` endpoint, specifically aggregation of geoJSON Features across flight-levels and/or time and/or aircraft class.

Similarly, the netCDF response object is refactored to move the `forecast_reference_time` from a data-array-wide attribute to a non-dimension coordinate, effectively allowing for a future case where multiple slices of `time` are returned, each with a difference corresponding `forecast_reference_time`.

## Refs

These contractual changes are staged for deployment in the **api-preprocessor**, [PR HERE](https://github.com/contrailcirrus/api-preprocessor/pull/77) and in the **contrails-api** [PR HERE](https://github.com/contrailcirrus/contrails-api/pull/152).

### CODE REVIEW

File: docs/specs/forecast-api.md [hunk 1]
```diff
@@ -20,7 +20,9 @@ availability.
     hour interval, at all longitudes and latitudes ranging from \[-80,
     80\] (this is expected to expand and encompass the poles, with an
     update at a later time). These grids are available across several
-    [aircraft classes](#aircraft-classes).  The variable returned in the netCDF
+    [aircraft classes](#aircraft-classes); only the `default` aircraft class is served until 
+    there is sufficient customer evidence to suggest needing the implementation of additional aircraft classes.
```
Comment (mlshapiro) [inline docs/specs/forecast-api.md]:
```suggestion
    there is sufficient evidence and agreement to implement additional aircraft classes. 
```

File: docs/specs/forecast-api.md [hunk 2]
```diff
@@ -34,15 +36,15 @@ availability.
...
-machine-to-machine integration. These two endpoints represent a minimal
+These two endpoints represent a minimal
 yet complete set of capabilities for air traffic planners to explore and
 prototype contrail avoidance in air traffic decision support systems.
 
 The signature, behavior and design considerations of these two endpoints
 are detailed below.
 
-Future work may include adding a discovery/recommendation endpoint, to
+Future work, contingent upon the implementation of multiple aircraft classes,
```
Comment (mlshapiro) [inline docs/specs/forecast-api.md]:
```suggestion
If/when multiple aircraft classes are implemented, future work may include a discovery/recommendation endpoint to
```

Comment (mlshapiro) [reply docs/specs/forecast-api.md]: (semantic suggestion that isn't necessarily better, feel free to reject)

File: docs/specs/forecast-api.md [hunk 3]
```diff
@@ -108,7 +105,8 @@ and ordinal yet leave sufficient room for future expansion.
 
 ### timestamp
 
-The `ts` value has the form `%Y%m%d%H`. This hour-resolution timestamp
+The `ts` value has the form `%Y%m%d%H`. It is a required query parameter.
```
Comment (mlshapiro) [inline docs/specs/forecast-api.md]:
```suggestion
The `{ts}` value has the form `%Y%m%d%H`. It is a required query parameter.
```

File: docs/specs/forecast-api.md [hunk 4]
```diff
@@ -123,7 +121,7 @@ generated using the nearest forecast time.
 ### flight level
 
 The fl value represents the flight level (*hectofeet*) of the returned
```
Comment (mlshapiro) [inline docs/specs/forecast-api.md]: The `{fl}` value represents the flight level (*hectofeet*) of the returned

File: docs/specs/forecast-api.md [hunk 5]
```diff
@@ -176,14 +160,14 @@ Coordinates:
...
 Data variables:
     contrails   (longitude, latitude, flight_level, time) float32 4MB ...
 Attributes:
-    forecast_reference_time:  "2024-07-01T06:00:00Z"
     aircraft_class: "default"
 ```
 
-Note that `forecast_reference_time` in the dataset level *Attributes* has the same definition as
+Note that `forecast_reference_time` non-dimension coordinate has the same definition as
 `model_run_at` in the API response header.
```
Comment (mlshapiro) [inline docs/specs/forecast-api.md]: Are we still including `model_run_at`? Is this comment necessary anymore?

File: docs/specs/forecast-api.md [hunk 6]
```diff
@@ -277,7 +260,15 @@ format:
   "features": [
       {
         "type": "Feature",
-        "properties": {},
+        "properties": {
+            "aircraft_class": "default",
+            "timestamp": "<%Y-%m-%dT%H:%M:%S>",
+            "flight_level": <int>,
+            "threshold": <int>,
+            "model_forecast_hour": <int>,
```
Comment (mlshapiro) [inline docs/specs/forecast-api.md]: Are we including any `model_...` properties anymore?

My suggestion would be to just include the `forecast_reference_time` only to be consistent with the netCDF file.  If you want to include `forecast_period` (instead of `model_forecast_hour`) this would be the integer hours between `timestamp` and `forecast_reference_time`

Or if you want to include these, then copy the definitions from up above down here. I would only include

Rather just the `forecast_reference_time` the same as in the netCDF file?

**Review by mlshapiro — APPROVED** [review]
This looks good - only comments about consistency in what was formerly Response Headers (`model_...`) and now shows up in JSON properties.

File: docs/specs/forecast-api.md (hunk 4 above)
Comment (nickmasson) [inline docs/specs/forecast-api.md]: consistency fix:
```suggestion
The {fl} value represents the flight level (*hectofeet*) of the returned
```

File: docs/specs/forecast-api.md (hunk 5 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md]: I don't think so (and agreeing w/ you suggestion below to just include `forecast_reference_time` in the geoJSON properties).

We can strike it here (and elsewhere), thereby making the `model_*` quantities something of an internal definition/concern only.

File: docs/specs/forecast-api.md (hunk 6 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md]: I agree.

I think we can strike the `model_*` quantities (for two reasons).

1. nice to make this nomenclature an internal concern only
2. anybody concerning themselves with model run at vs. predicted at... will be well enough informed to know that the `"timestamp"` quantity here is equivalent to `"predicted_at"`, `"forecast_reference_time"` is equiv. to `"model_run_at"`, and `"model_forecast_hour"` is the diff of the two.

File: docs/specs/forecast-api.md [hunk 7]
```diff
@@ -122,8 +120,8 @@ generated using the nearest forecast time.
 
 ### flight level
 
-The fl value represents the flight level (*hectofeet*) of the returned
-netCDF global grid.
+The {fl} value represents the flight level (*hectofeet*) of the returned
```
Comment (mlshapiro) [inline docs/specs/forecast-api.md]: Shouldn't the {fl} be in a code block, i.e. `{fl}`?

Comment (nickmasson) [reply docs/specs/forecast-api.md]: yes; inline. I'll also scan for other similar fmt fixes

File: docs/specs/forecast-api.md [hunk 8]
```diff
@@ -176,16 +160,13 @@ Coordinates:
   * latitude   (latitude) float32 3kB -80.0 -79.75 -79.5 ... 79.5 79.75 80.0
   * flight_level      (flight_level) int16 2B 300
   * time       (time) datetime64[ns] 8B 2024-07-01T12:00:00
+    forecast_reference_time (time) datetime64[ns] 8B 2024-07-01T06:00:00
```
Comment (max-vogler) [inline docs/specs/forecast-api.md:173]: A tiny question for `time` and `forecast_reference_time`: Is it intentional to remove the timezone `Z` from the otherwise ISO8601 timestamp?

**Review by max-vogler — APPROVED** [review]
The changes generally look good! Thanks for working with us on a flexible API call signature.

The main future change I anticipate from our end is to make `fl` and `threshold` optional to allow returning grids/regions for all flight levels and thresholds. These changes will be backwards compatible with the proposed design here, so all good!

File: docs/specs/forecast-api.md (hunk 8 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md:173]: great question. I believe time-like data is actually stored as int64 in the netCDF -- the types here reflect the marshaling that xarray does when importing the file, so the timezone naive type here is auto-magically set by xarray.

@mlshapiro can you check me on correctness here?

Comment (nickmasson) [reply docs/specs/forecast-api.md:173]: That does bring up a good questions as to whether we'd want to suffix `Z` on the timestamps returned in the geoJSON properties, however: https://github.com/contrailcirrus/contrail-forecast/blob/refactor/v1-signature/docs/specs/forecast-api.md#response-object-1

As it is now, we have consistency. Timezone naive timestamps throughout; implicit UTC.

Adding suffixes in the Propeties of the geoJSON may create mixed conventions, but add specificity to the geoJSON response, at least.

I could see arguments for either way. But I leaning towards consistency (and of course, there is always documentation).

Comment (nickmasson) [reply docs/specs/forecast-api.md:173]: Also noting that the `timestamp` passed as a query param (`timestamp="%Y%m%d%H"`) is timezone naive, UTC implicit as well.

If we wanted to be timezone aware here... we'd probably want to go to `timestamp=<int>` unixtime or `timestamp=<iso_datetime_str>`.  I see that being more canonical in general... but we loose the self-documenting goodness of the current design (which communicates to customers that response data only exists on the hour).

Comment (max-vogler) [reply docs/specs/forecast-api.md:173]: Could you change the signature to `time=ISO 8601`? Renaming to `time` would match both the coordinate in your return data and our parameter/coordinate name. I'd personally prefer explicit documentation around the coarseness of times, while allowing any ISO 8601 time to be passed in. We might change the forecast to exist more often than only once per hour in the future.

File: docs/specs/forecast-api.md [hunk 9]
```diff
@@ -122,8 +120,8 @@ generated using the nearest forecast time.
 
 ### flight level
 
-The fl value represents the flight level (*hectofeet*) of the returned
-netCDF global grid.
+The `{fl}` value represents the flight level (*hectofeet*) of the returned
```
Comment (max-vogler) [inline docs/specs/forecast-api.md:134]: Could you rename this to `flight_level`? This would match both the result coordinate name and our parameter/coordinate name.

File: docs/specs/forecast-api.md [hunk 10]
```diff
@@ -176,16 +160,13 @@ Coordinates:
   * latitude   (latitude) float32 3kB -80.0 -79.75 -79.5 ... 79.5 79.75 80.0
   * flight_level      (flight_level) int16 2B 300
   * time       (time) datetime64[ns] 8B 2024-07-01T12:00:00
+    forecast_reference_time (time) datetime64[ns] 8B 2024-07-01T06:00:00
 Data variables:
     contrails   (longitude, latitude, flight_level, time) float32 4MB ...
 Attributes:
-    forecast_reference_time:  "2024-07-01T06:00:00Z"
```
Comment (max-vogler) [inline docs/specs/forecast-api.md:182]: qq: What's the reason for migrating `forecast_reference_time` to be a coordinate and no longer an attribute? We have a similar attribute today and would then evaluate the same change!

File: docs/specs/forecast-api.md (hunk 8 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md:173]: > We might change the forecast to exist more often than only once per hour in the future.

👍 

> Could you change the signature to time=ISO 8601? Renaming to time would match both the coordinate in your return data and our parameter/coordinate name. I'd personally prefer explicit documentation around the coarseness of times, while allowing any ISO 8601 time to be passed in.

Completely agreed on consistency of naming (`timestamp` -> `time`).

On field definition loosening to any iso8601 date(time) str..., we loose the benefits of the quantity being self-describing which is a downside , and generally loosen product-fit. But I see the motivation given the general trend here towards maintaining flexibility at the expensive of product fit.  Given that you'll be owning this long-term, I see no problem in updating to accept any iso8601 string.

If we make that change, I see two behavioral considerations:
1) what would be your preferred behavior if the provided value doesn't match an available prediction? For instance, predictions are currently only available hourly (aligned w/ frequency of meteorological forecast availability).  If a caller provides `"2024-01-02T01:12:00"`, would you prefer we snap to the nearest available forecast, or return a 404, or other?

2) do you want to support timezone handling? should the tz designator be required? If not required, then implicit UTC?

File: docs/specs/forecast-api.md (hunk 10 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md:182]: If you want to commit to only ever serving a single time slice, then I think having it as an attr makes sense. If you want to option available to return multiple time slices, then it should move to a coord (that references `time` 1:1).

--

The `forecast_reference_time` is the time at which the _meteorological model_ (used in generating the contrails predictions) was executed for a given `time`.  So multiple time slices may have different `forecast_reference_time` values.

Possible alternatives I can think of, but that seem less favorable:
- support multiple times in query but return multiple netcdfs for a given call, keeping `forecast_reference_time` as an attr
- support multiple times in query and concat slices into a single netcdf, but have all time slices reference a single `forecast_reference_time` (this would (1) limit the valid time slice candidates that could be concat'ed into a single file, and, (2) force choosing `time` slices that share a common `forecast_reference_time` instead of the latest available `forecast_reference_time` for a given `time`.

--
FWIW, I think its a good idea to stay away from supporting multiple time slices or time ranges in a query. It would make adding a 305 response option tricky in the future (many customers we've talked to expect to use the API as part of a caching service, so a 305 would be helpful).

File: docs/specs/forecast-api.md (hunk 9 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md:134]: You mean the param name, right? If so, I think we're already good (see signature here):
https://github.com/contrailcirrus/contrail-forecast/blob/refactor/v1-signature/docs/specs/forecast-api.md#signature-1

The bracketed fields just represent the parameter value passed.
We're just using shorthand for brevity in the signature documentation, but if you'd prefer that the signature read:
`?flight_level={flight_level}&time={time}` etc... instead of `?flight_level={fl}&time={ts}`, I'm OK with that.

File: docs/specs/forecast-api.md (hunk 10 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md:182]: pinning this here as it is a relevant reminder for context ⏫ 

https://github.com/contrailcirrus/contrail-forecast/blob/refactor/v1-signature/docs/specs/forecast-api.md#timestamp

> Given that ECMWF delivers 73 hours of forecast data every 6 hours, it is expected that a given ts will have multiple candidate URIs to serve to a client.

> If multiple candidate URIs exist for a given timestamp at the time of the client request, the API should return the URI whose outputs were generated *using the nearest forecast time*.

File: docs/specs/forecast-api.md [hunk 11]
```diff
@@ -277,7 +257,13 @@ format:
   "features": [
       {
         "type": "Feature",
-        "properties": {},
+        "properties": {
+            "aircraft_class": "default",
+            "timestamp": "<%Y-%m-%dT%H:%M:%S>",
```
Comment (nickmasson) [inline docs/specs/forecast-api.md]: probably makes sense to suffix `Z` here  for avoidance of doubt (`"<%Y-%m-%dT%H:%M:%SZ>"`), if there is a future possibility of accepting timezone aware strings in the `time` query param.

File: docs/specs/forecast-api.md [hunk 12]
```diff
@@ -277,7 +257,13 @@ format:
...
+            "flight_level": <int>,
+            "threshold": <int>,
+            "forecast_reference_time": <%Y-%m-%dT%H:%M:%S>
```
Comment (nickmasson) [inline docs/specs/forecast-api.md]: probably makes sense to suffix Z here for avoidance of doubt ("<%Y-%m-%dT%H:%M:%SZ>"), if there is a future possibility of accepting timezone aware strings in the time query param.

File: docs/specs/forecast-api.md (hunk 11 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md]: this will also be renamed to `time`

**Comment by nickmasson** [comment]
@max-vogler, consolidating action items from our above conversations here. Please let me know if this resolves all your requests above, and I will resolve those conversations and make these changes:

(1) `/regions` and `/grids` path will both accept a `time` query param (renaming `timestamp`)

(2) `time` query param's value must be a valid iso 8601 date(time) string
(2a) exact time matches only: if `time` value does not exactly match an available contrails prediction (hourly interval), e.g. `"2024-01-01T12:04:15"`, caller will receive a 404
(2b) only timezone naive iso8601 supported: if `time` value has a timezone designator suffix, caller will receive a 404 

(3)`timestamp` parameter in geoJSON Feature will be renamed to `time`
(3a) `time` (fmrly `timestamp`) param in geoJSON Feature will receive suffix `Z`
(3b) `forecast_reference_time` param in geoJSON Feature will receive suffix `Z`

Does this work for you? (note: no problem if you want to relax 2a/b in your implementation, but this is what I propose for our current deploy).

File: docs/specs/forecast-api.md (hunk 9 above)
Comment (max-vogler) [reply docs/specs/forecast-api.md:134]: This was an oversight from my end – the short form for the placeholder is perfectly fine! Thanks for pointing it out :)

File: docs/specs/forecast-api.md (hunk 8 above)
Comment (max-vogler) [reply docs/specs/forecast-api.md:173]: 1. Snapping to the nearest forecast would be perfect. This would allow us to be flexible in what timestamps we accept which IMO makes the API easily usable. Callers can then learn which actual forecast was provided in the response through the reference timestamp dimension.

2. No strong opinion, your proposal sounds good.

File: docs/specs/forecast-api.md (hunk 10 above)
Comment (max-vogler) [reply docs/specs/forecast-api.md:182]: Thanks for elaborating, this really helped my understanding!

1) What makes caching a NetCDF response that has multiple concatenated time slices harder than caching a single time slice? Anticipating multiple time slices might otherwise be great API design! :)

2) Would it be backwards compatible for us to add the new reference time dimension to the existing NetCDF response, or is adding a new dimension generally considered breaking?

**Comment by max-vogler** [comment]
Thanks for the in-depth reply! 1 & 3 SG, for 2 I'd propose to be flexible in what we accept – snapping to the closest matching timestamp and accepting both timezone naive (interpreted as UTC) and timezone-aware timestamps.

File: docs/specs/forecast-api.md (hunk 10 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md:182]: > What makes caching a NetCDF response that has multiple concatenated time slices harder than caching a single time slice? Anticipating multiple time slices might otherwise be great API design! :)

TL;DR -- if we have N time slices, it is very possible that only a few slices are modified at any given time and warrant cache updates.  Sending N slices over the wire every time is sub-optimal.  Providing change detection (modification) on a per time slice basis, and enforcing retrieval on a per-time-slice basis, more strongly guarantees a client will have quality caching behavior (albeit with a bit more complexity to manage concurrency).

Longer answer:
If the contrail predictions are using ECMWF meteorological forecasts, then there will be (nominally) 12 different met forecasts for a given target `time`.  (ECMWF provides forecasts 73hrs into the future, now() inclusive, and they re-run the forecast every 6hrs, so any given hourly `time` has 73/6 overlapping forecast options).

As such, a query for a given `?time=splat` can return 12 different contrail forecasts, depending on the time of the call (default behavior, as per customer request and expected usage, being to return the latest forecast available).  This is why there is a note about the API not being idempotent. This is akin to checking for Friday's weather today, then checking for Friday's weather again tomorrow, and expecting the weather values to be different.  Noting that a future "super-user" upgrade could certainly be to accept a `?forecast_time=splat` param, which would guarantee idempotent behavior, but make client integration logic considerably more complex.

So to the question of caching logic wrt multiple time slices, a given netCDF file would have 12**N variants given N `time` slices in the netCDF.  A 305 response (or other similar change detection mechanism) would trigger whenever _any_ of the N slices is modified.  If a client is well informed wrt how the forecasts are due to update, they could get a high hit-rate of modified slides (N_mod/N_tot ~=1), but, making this assumption seems antithetical to a flexible API design (separating domain logic concerns from integration logic concerns). 

This is a bit sticky and I can see a few other ways that this could have been designed, but I won't elaborate here (this design was chosen given our opinions on how flight planners will want machine-2-machine integration without a big integration logic burden).  I'm happy to jump on a call and discuss the time/forecast_time behavior in more depth if you're interested.

> Would it be backwards compatible for us to add the new reference time dimension to the existing NetCDF response, or is adding a new dimension generally considered breaking?

Great question. I actually don't know from a netCDF consumer perspective. I'll inquire w/ our team today as they have more experience working w/ netCDF using various client-side tools. 

I _would_ expect that having the `forecast_time` as an attr, and then removing it (migrating to a coord) would be breaking.  So perhaps we keep it as a coord to start? (I don't see other substantive downside other than that `forecast_time` "feels" more like an attr than a first-class coord).

**Comment by nickmasson** [comment]
> Thanks for the in-depth reply! 1 & 3 SG, for 2 I'd propose to be flexible in what we accept – snapping to the closest matching timestamp and accepting both timezone naive (interpreted as UTC) and timezone-aware timestamps.

I'm good with these. I'll update this document as such and reflect the changes in the next release.

@max-vogler I think the only remaining conversation to resolve is the one regarding `forecast_time` as an attr or coord in the netCDF, and any open threads that need tying up on that conversation.

File: docs/specs/forecast-api.md (hunk 10 above)
Comment (nickmasson) [reply docs/specs/forecast-api.md:182]: following up here.
It shouldn't be breaking to add a new non-indexed coordinate to the netCDF at a future time.

It is important, however, to include `forcast_reference_time`... and if we include it in an attribute and then decide to migrate to a non-indexed coord later, it would be a breaking change to remove the _attribute_ in favor of the new non-indexed coordinate.

So if there is any possibility of wanting to return multiple `time` slices in the future, it seems reasonable to just have `forecast_reference_time` be a non-indexed coord (or, we can agree that we won't leave the back door open for aggregations on `time` and leave `forecast_reference_time` an attr... ◀️  my preference).

**Comment by max-vogler** [comment]
Thanks for the in-depth explanation and checking back with the team.  Slight preference for a non-indexed coordinate instead of an attribute from my side – seems to accommodate a lot of potential future changes we could think of! If you have  strong preference for the attribute, this is fine from our side as well. LGTM in general :)

File: docs/specs/forecast-data.md [hunk 13]
```diff
@@ -28,8 +24,10 @@ Additional attributes, in addition to the required and suggested ones above, may
 - `longitude` (`float32`): `np.arange(-180, 180, 0.25)`, EPSG:4326
 - `latitude` (`float32`): `np.arange(-90, 90, 0.25)`, EPSG:4326
 - `flight_level` (`int16` or `int32`): `[270, 280, 290, 300, 310, 320, 330, 340, 350, 360, 370, 380, 390, 400, 410, 420, 430, 440]`, hectofeet [^flightlevels]
-- `time` (`int32` or `int64`): [CF compatible time coordinates](https://cfconventions.org/cf-conventions/cf-conventions#time-coordinate). Must have `units` and `calendar` variable attributes. e.g.
+- `time` (`int32` or `int64`): [CF compatible time coordinates](https://cfconventions.org/cf-conventions/cf-conventions#time-coordinate).
+- `forecast_reference_time` (`int32` or `int64`): The forecast reference time is the "data time", i.e. the time at which the meteorological model was executed for a given set of forecast times.`
 
+💡 Both `time` and `forecast_reference_time` must have `units` and `calendar` variable attributes. e.g.
```
Comment (nickmasson) [inline docs/specs/forecast-data.md]: @mlshapiro & @zebengberg, I think this section of the data contract could use more elaboration. Up to you to decide how best to phrase...

My concern is that this requirement, as phrased, is a bit confusing.
xarray, when reading in a netCDF, will _not_ show a `units` or `calendar` attr on `time`.  

In inspecting an existing grid netcdf from our v0 API, I see that these attrs are set when inspecting the file with `ncdump`.

When loaded with xarray, however, the attrs are scrubbed from `time`. I see it is automagically set when serializing to netCDF, and deserializing from netCDF.

Furthermore, the `units` reference value is not static; it is dynamically set, which is confusing for folks who are used to e.g. epoch time with a static reference of 1970-01-01...

Comment (mlshapiro) [reply docs/specs/forecast-data.md]: It is written confusingly, but this is how its written in the [CF Convention](https://cfconventions.org/cf-conventions/cf-conventions#time-coordinate)

I will re-arrange this section to hopefully be a little more clear.

While we use xarray for everything, many others won't, so it should be true to the CF convention. Meterologists will be very comfortable with this format, but its true that developers will not be as versed in CF vagaries...

Comment (mlshapiro) [reply docs/specs/forecast-data.md]: For future reference, the CF convention could support unix time if desired. In the `xarray` paradigm, you'd convert the time coordinate to an `int64` representing seconds since 1970-01-01 00:00:00 and then write a `units` attr that days `"seconds since 1970-01-01 00:00:00"`, e.g.

```
ds["time"] = ((ds["time"] - np.datetime64('1970-01-01T00:00:00')) / np.timedelta64(1, 's')).astype(np.int64)
ds["time"].attrs["units"] = "seconds since 1970-01-01 00:00:00"
```

Then write that out. Calendar is still relevant, so its worth saying its required.

Comment (mlshapiro) [reply docs/specs/forecast-data.md]: Updated the [Dimensions](https://github.com/contrailcirrus/contrail-forecast/blob/refactor/v1-signature/docs/specs/forecast-data.md#dimensions) section and added a little [time dimensions](https://github.com/contrailcirrus/contrail-forecast/blob/refactor/v1-signature/docs/specs/forecast-data.md#time-dimensions) explainer. Resolve if satisfied!

Comment (nickmasson) [reply docs/specs/forecast-data.md]: clear ty!

could maybe use a tldr one-liner at the head of our diff along the lines of:
```
Xarray will automatically include a `units` and `calendar` attribute on all time-like variables when writing to a netCDF file.
{details here}
```
